dashboard/lib/services/user/upgrade_to_personal_login.rb

```
@@ -0,0 +1,46 @@
+module Services
+  module User
+    class UpgradeToPersonalLogin < Services::Base
+      attr_reader :user, :params
+
+      def initialize(user:, params:)
+        @user = user
+        @params = params.dup
+      end
+
+      def call
+        return false unless user.student?
+
+        if user.secret_word_account? && !user.valid_secret_words?(params[:secret_words])
+          error = params[:secret_words].blank? ? :blank_plural : :invalid_plural
+          user.errors.add(:secret_words, error)
+          return false
+        end
+
+        unless user.migrated?
+          params[:provider] = nil
+          return user.update(params.except(:secret_words))
```

Is the existing code unnecessarily updating the `:secret_words` column, and this fixes it by excluding it from the `user.update` call? Same with below on line 27.

---

That's a good question.
I am unsure. To me, it looks like `secret_words` was used for validation, but the value is not updated at all.

I skipped updatign the field because writing it again is unnecessary. Worse, if someone tampers with the request and sends a new (incorrect) `secret_words`, we would end up silently replacing the correct words.

Do you know if personal login accounts have secret words or do they get removed? I quickly scanned the user model and didn't see any callbacks or methods removing secret words other than when purging a user.

---

We only assign secret pictures/words to sponsored accounts https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/user.rb#L376-L379

This was changed a while ago in [this PR](https://github.com/code-dot-org/code-dot-org/pull/62549/files#diff-0a5e79b422e8d2a2b8e5d9cc2203f62899ee74b65da749306b3d5f7c8b21a2d2R545-R548)